File: src/backend/reference-api.mjs
import { createServer } from 'node:http';
import { readFile } from 'node:fs/promises';
import { fileURLToPath } from 'node:url';
import path from 'node:path';

const schemaPath = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '../../references/ballistics-reference-schema.json');
const port = Number(process.env.ARTY_REFERENCE_PORT || 8090);

const schema = JSON.parse(await readFile(schemaPath, 'utf8'));

function json(res, status, body) {
  res.writeHead(status, { 'content-type': 'application/json; charset=utf-8', 'cache-control': 'no-store' });
  res.end(JSON.stringify(body));
}

function validatePayload(payload) {
  const errors = [];
  if (!payload || typeof payload !== 'object' || Array.isArray(payload)) errors.push('payload ต้องเป็น object');
  if (payload?.operationalCalculationEnabled === true) errors.push('ไม่อนุญาตให้เปิดการคำนวณเชิงปฏิบัติการ');
  for (const field of schema.restrictedPayloadPolicy.blockedFields) {
    if (payload && Object.prototype.hasOwnProperty.call(payload, field)) errors.push(`ไม่อนุญาต field: ${field}`);
  }
  return { valid: errors.length === 0, errors, mode: 'schema-validation-only' };
}

const server = createServer(async (req, res) => {
  const url = new URL(req.url ?? '/', `http://${req.headers.host ?? 'localhost'}`);
  if (req.method === 'GET' && url.pathname === '/api/v1/health') {
    return json(res, 200, { status: 'ONLINE', service: 'arty-reference-api', mode: 'schema-validation-only', operationalCalculationEnabled: false });
  }
  if (req.method === 'GET' && url.pathname === '/api/v1/reference/schema') return json(res, 200, schema);
  if (req.method === 'GET' && url.pathname === '/api/v1/reference/tables') return json(res, 200, { tables: schema.tableCatalog });
  if (req.method === 'POST' && url.pathname === '/api/v1/reference/validate') {
    let body = '';
    req.setEncoding('utf8');
    req.on('data', (chunk) => { body += chunk; });
    req.on('end', () => {
      try { return json(res, 200, validatePayload(JSON.parse(body || '{}'))); }
      catch { return json(res, 400, { valid: false, errors: ['JSON ไม่ถูกต้อง'] }); }
    });
    return;
  }
  return json(res, 404, { error: 'ไม่พบ endpoint' });
});

server.listen(port, () => console.log(`ARTY reference API listening on ${port}`));

File: src/backend/reference_api.py
"""ARTY V.2 reference-only API.

Optional FastAPI adapter. It validates provenance/schema payloads only and
deliberately does not expose firing tables, ballistic solutions, or commands.
"""
import json
import os
from pathlib import Path
from typing import Any

from fastapi import FastAPI, HTTPException

SCHEMA_PATH = Path(__file__).resolve().parents[2] / "references" / "ballistics-reference-schema.json"
SCHEMA = json.loads(SCHEMA_PATH.read_text(encoding="utf-8"))
app = FastAPI(title="ARTY V.2 Reference API", version="1.0.0")

@app.get("/api/v1/health")
def health() -> dict[str, Any]:
    return {"status": "ONLINE", "service": "arty-reference-api", "mode": "schema-validation-only", "operationalCalculationEnabled": False}

@app.get("/api/v1/reference/schema")
def schema() -> dict[str, Any]:
    return SCHEMA

@app.get("/api/v1/reference/tables")
def tables() -> dict[str, Any]:
    return {"tables": SCHEMA["tableCatalog"]}

@app.post("/api/v1/reference/validate")
def validate(payload: dict[str, Any]) -> dict[str, Any]:
    blocked = SCHEMA["restrictedPayloadPolicy"]["blockedFields"]
    errors = []
    if payload.get("operationalCalculationEnabled") is True:
        errors.append("ไม่อนุญาตให้เปิดการคำนวณเชิงปฏิบัติการ")
    errors.extend(f"ไม่อนุญาต field: {field}" for field in blocked if field in payload)
    return {"valid": not errors, "errors": errors, "mode": "schema-validation-only"}

if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="127.0.0.1", port=int(os.getenv("ARTY_REFERENCE_PORT", "8090")))
